text/0000-inline-asm.md

```diff
@@ -0,0 +1,918 @@
+
+Take the following (AArch64) asm block as an example:
+
+```rust
+unsafe fn foo(mut a: i32, b: i32) -> (i32, i32)
+{
+    let c;
+    asm!("<some asm code>", inout(reg) a, in("x0") b, out("x20") c);
+    (a, c)
+}
+```
+
+This could be expanded to an external asm file with the following contents:
+
+```
+# Function prefix directives
+.section ".text.foo_inline_asm"
+.globl foo_inline_asm
+.p2align 2
+.type foo_inline_asm, @function
+foo_inline_asm:
+
+// If necessary, save callee-saved registers to the stack here.
+str x20, [sp, #-16]!
+
+// Move the pointer to the argument out of the way since x0 is used.
+mov x1, x0
+
+// Load inputs values
+ldr w2, [x1, #0]
+ldr w0, [x1, #4]
+
+<some asm code>
+
+// Store output values
+str w2, [x1, #0]
+str w20, [x1, #8]
+
+// If necessary, restore callee-saved registers here.
+ldr x20, [sp], #16
+
+ret
+
+# Function suffix directives
+.size foo_inline_asm, . - foo_inline_asm
+```
+
+And the following Rust code:
+
+```rust
+unsafe fn foo(mut a: i32, b: i32) -> (i32, i32)
+{
+    let c;
+    {
+        #[repr(C)]
+        struct foo_inline_asm_args {
+            a: i32,
+            b: i32,
+            c: i32,
+        }
+        extern "C" {
+            fn foo_inline_asm(args: *mut foo_inline_asm_args);
+        }
+        let mut args = foo_inline_asm_args {
+            a: a,
+            b: b,
+            c: mem::uninitialized(),
+        };
+        foo_inline_asm(&mut args);
+        a = args.a;
+        c = args.c;
+    }
+    (a, c)
+}
+```
+
+## Rules for inline assembly
+[rules]: #rules
+
+- Any registers not specified as inputs will contain an undefined value on entry to the asm block.
+- Any registers not specified as outputs must have the same value upon exiting the asm block as they had on entry.
+  - This only applies to registers which can be specified as an input or output.
+- Behavior is undefined if execution unwinds out of an asm block.
+- Any memory reads/writes performed by the asm code follow the same rules as `volatile_read` and `volatile_write`.
+  - Refer to the unsafe code guidelines for the exact rules.
+  - If the `readonly` option is set, then only memory reads (with the same rules as `volatile_read`) are allowed.
+  - If the `nomem` option is set then no reads or write to memory are allowed.
+- Unless the `nostack` option is set, asm code is allowed to use stack space below the stack pointer.
+  - On entry to the asm block the stack pointer is guaranteed to be suitably aligned (according to the target ABI) for a function call.
+  - You are responsible for making sure you don't overflow the stack (e.g. use stack probing to ensure you hit a guard page).
```

Comment:
I think this should be carefully considered. It seems like this places an unnecessary burden on the user, especially if you don’t have control/knowledge of whatever stack protection Rust already decides to use. Also, since you cannot take advantage of whatever checks Rust already does, attempting to repeat such checks yourself would be a huge performance hit, but doing so could be necessary. Otherwise for example, what if you are using an embedded MCU with no stack protection or heap, create a large local array, and the array index overflows (somewhat feasible for 32-bit)? Do you need to check for overflow yourself at the start of each asm block, even though Rust may already check that there is enough space for what it knows about the stack frame (e.g. Rust local variables)?

One solution would be adding an option to reserve some amount of stack space. For example `stack(16)` (or with a better name) would reserve 16 bytes directly below the stack for you to use, i.e. at `esp - 0x10`.